src/dftbp/dftb/repulsive/chimesrep.F90

```diff
@@ -0,0 +1,317 @@
+    this%stress(:,:) = this%stress * (kcal_mol__Hartree / AA__Bohr**3)
+
+  end subroutine TChimesRep_updateCoords
+
+
+  !> Returns the energy of the repulsive.
+  subroutine TChimesRep_getEnergy(this, coords, species, img2CentCell, neighbourList, Eatom,&
+      & Etotal, iAtInCentralRegion)
+
+    !> Instance.
+    class(TChimesRep), intent(in) :: this
+
+    !> All atomic coordinates
+    real(dp), intent(in) :: coords(:,:)
+
+    !> All atoms chemical species
+    integer, intent(in) :: species(:)
+
+    !> Image atom indices to central cell atoms
+    integer, intent(in) :: img2CentCell(:)
+
+    !> List of neighbours for each atom
+    type(TNeighbourList), intent(in) :: neighbourList
+
+    !> Energy for each atom
+    real(dp), intent(out) :: Eatom(:)
+
+    !> Total energy
+    real(dp), intent(out) :: Etotal
+
+    !> Atoms in the central cell (or device region if transport)
+    integer, intent(in), optional :: iAtInCentralRegion(:)
+
+    if (present(iAtInCentralRegion)) then
+      Eatom(:) = 0.0_dp
+      Eatom(iAtInCentralRegion) = this%energyPerAtom(iAtInCentralRegion)
+    else
+      Eatom(:) = this%energyPerAtom
+    end if
+    Etotal = sum(Eatom)
+
+  end subroutine TChimesRep_getEnergy
+
+
+  !> Returns the gradients of the repulsive interaction
+  subroutine TChimesRep_getGradients(this, coords, species, img2CentCell, neighbourList, grads)
+
+    !> Instance
+    class(TChimesRep), intent(in) :: this
+
+    !> coordinates (x,y,z, all atoms including possible images)
+    real(dp), intent(in) :: coords(:,:)
+
+    !> Species of atoms in the central cell.
+    integer, intent(in) :: species(:)
+
+    !> Index of each atom in the central cell which the atom is mapped on.
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Neighbour list
+    type(TNeighbourList), intent(in) :: neighbourList
+
+    !> Gradient for each atom. Shape: [3, nAtom]
+    real(dp), intent(out) :: grads(:,:)
+
+    grads(:,:) = this%gradients
+
+  end subroutine TChimesRep_getGradients
+
+
+  !> Returns the stress tensor contribution from the repulsive term
+  subroutine TChimesRep_getStress(this, coords, species, img2CentCell, neighbourList, cellVol,&
+      & stress)
+
+    !> Instance
+    class(TChimesRep), intent(in) :: this
+
+    !> coordinates (x,y,z, all atoms including possible images)
+    real(dp), intent(in) :: coords(:,:)
+
+    !> Species of atoms in the central cell.
+    integer, intent(in) :: species(:)
+
+    !> indexing array for periodic image atoms
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Neighbour list
+    type(TNeighbourList), intent(in) :: neighbourList
+
+    !> cell volume.
+    real(dp), intent(in) :: cellVol
+
+    !> stress tensor
+    real(dp), intent(out) :: stress(:,:)
+
+    stress(:,:) = this%stress
+
+  end subroutine TChimesRep_getStress
+
+#:else
+
+  !> Dummy initializer in case code was compiled without ChIMES
+  subroutine TChimesRep_init()
+  end subroutine TChimesRep_init
```

Comment:
Similarly, just remove the body of the procedures and replace it with an error instead if having a separate stub?

Reply:
This is open for discussion. I can see following advantages, if we only provide names without content (empty type, dummy subroutine without args):
* we do not need to guard the module imports by preprocessor instructions, as the name exists, 
* we are forced, however, to guard all places, where the non-existing types and subroutines are actually used, by `#:if WITH_*` instructions. Forgetting such a guard would cause a compile time error (as opposed to a run-time error, we would get by using dummies with the same public interface as the original ones),
* coverage won't decrease due to all the `error stop`s in the dummy routines, which are (hopefully) never executed during the tests.